cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,387 @@
+		l.logAndFatalf("Abort. Upgrade is currently not supported: %v -> %v.", cur, tar)
+	}
+
+	l.logAndPrintf("Version check passed: %v -> %v.\n", cur, tar)
+}
+
+func parseVersionFormat(ver string) (int, int, int) {
+	fullVerArray := strings.Split(ver, "-")
+	if len(fullVerArray) == 0 {
+		l.logAndFatalf("Abort. Incorrect version: %v.", ver)
+	}
+	verArray := strings.Split(fullVerArray[0], ".")
+	if len(verArray) != 3 {
+		l.logAndFatalf("Abort. Incorrect version: %v.", ver)
+	}
+	major, err := strconv.Atoi(verArray[0])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect marjor version: %v.", verArray[0])
+	}
+	minor, err := strconv.Atoi(verArray[1])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect minor version: %v.", verArray[1])
+	}
+	patch, err := strconv.Atoi(verArray[2])
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect patch version: %v.", verArray[2])
+	}
+	return major, minor, patch
+}
+
+func retrieveControlPlaneVersion(kubeClient kubernetes.ExecClient, istioNamespace string) string {
+	meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+	if e != nil {
+		l.logAndFatalf("Failed to retrieve Istio control plane version, error: %v", e)
+	}
+
+	if meshInfo == nil {
+		l.logAndFatalf("Istio control plane not found in namespace: %v", istioNamespace)
+	}
+
+	for _, remote := range *meshInfo {
+		l.logAndPrintf("Control Plane - %s pod - version: %s", remote.Component, remote.Info.Version)
+	}
+
+	return coalesceVersions(meshInfo)
+}
+
+func waitUpgradeComplete(kubeClient kubernetes.ExecClient, istioNamespace string, targetVer string) {
+	l.logAndPrintf("Start waiting for upgrade rollout to complete...")
+	sleepSeconds(20)
+	for i := 1; i <= 60; i++ {
+		if i != 1 {
+			sleepSeconds(10)
+		}
+		l.logAndPrintf("Waiting for upgrade rollout to complete, attempt #%v: ", i)
+		meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+		if e != nil {
+			l.logAndPrintf("Failed to retrieve Istio control plane version, error: %v", e)
+			continue
+		}
+		if meshInfo == nil {
+			l.logAndPrintf("Failed to find Istio namespace: %v", istioNamespace)
+			continue
+		}
+		for _, remote := range *meshInfo {
+			if targetVer != remote.Info.Version {
+				l.logAndPrintf("Control Plane - %s pod - version %s does not match the target version %s",
+					remote.Component, remote.Info.Version, targetVer)
+			}
+		}
+		if identicalVersions(*meshInfo) {
+			l.logAndPrintf("Upgrade rollout completed. " +
+				"All Istio control plane pods are running with the same version.\n\n")
+			return
+		}
+		l.logAndPrint("Istio control plane pods are running on multiple versions.")
+	}
+	l.logAndFatal("Upgrade rollout unfinished. Maximum number of attempts exceeded, quit...")
+}
+
+func sleepSeconds(n int) {
+	l.logAndPrintf("Going to sleep for %v secounds ", n)
+	for i := 1; i <= n; i++ {
+		time.Sleep(time.Second)
+		fmt.Print(".")
+	}
+}
+
+func retrieveClientVersion() string {
+	l.logAndPrintf("Client - istioctl version: %s\n", istioVersion.Info.Version)
```

Comment:
It seems overkill to have a function for this.

Reply:
Removed.